### PR TITLE
To handle non-actionable steps in sklearn

### PR DESCRIPTION
**Pull request by Neeratyoy**
#### Reference Issue

Addresses #825 and #480. 

#### What does this PR implement/fix? Explain your changes.
The serialization expects each step to be a sklearn module and eventually an OpenMLFlow or a None. For ColumnTransform operations, a _drop_ or _passthrough_ can be specified to include a step in a pipeline but not execute it. The changes in this PR incorporates this additional string handling of the keywords _drop_ and _passthrough_ in a ColumnTransformer call.

#### How should this PR be tested?
```python
import openml
import numpy as np
import pandas as pd
from sklearn.pipeline import make_pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.tree import DecisionTreeClassifier

dataset_id = 61
dataset = openml.datasets.get_dataset(dataset_id)

task = openml.tasks.get_task(59)

X, y, categorical_ind, feature_names = dataset.get_data(target=dataset.default_target_attribute, dataset_format='array')
categorical_ind = np.array(categorical_ind)
cat_idx, = np.where(categorical_ind)
cont_idx, = np.where(~categorical_ind)

clf = make_pipeline(ColumnTransformer([('cat', make_pipeline(SimpleImputer(strategy='most_frequent'), OneHotEncoder()), cat_idx.tolist()),
                                       ('cont', make_pipeline(SimpleImputer(strategy='median'), StandardScaler()), cont_idx.tolist())])          
                , DecisionTreeClassifier(max_depth=1))
if not categorical_ind.any():
    clf[0].set_params(cat='drop')
if not (~categorical_ind).any():
    clf[0].set_params(cont='drop')
hotencoded = clf[:-1].fit_transform(X)

bla = openml.runs.run_model_on_task(model=clf, task=task, return_flow=True)
```

#### Any other comments?
This PR handles the specific case of a ColumnTransform. I'm not sure if there are other such keywords which should be handled equivalently. We can then maintain a keyword listing, based on the sklearn version under extension.py and handle them with a similar logic. 
If there are more diverse cases that may arise, some examples would help in designing a generalized solution in that case.
Would need to accordingly design relevant unit tests too.

### CODE REVIEW

File: openml/extensions/sklearn/extension.py [hunk 1]
```diff
@@ -694,10 +694,14 @@ def _serialize_model(self, model: Any) -> OpenMLFlow:
         # will be part of the name (in brackets)
         sub_components_names = ""
         for key in subcomponents:
+            if isinstance(subcomponents[key], str):
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Could you please add a comment why this is necessary?

File: openml/extensions/sklearn/extension.py [hunk 2]
```diff
@@ -769,21 +773,25 @@ def _get_external_version_string(
         external_versions.add(openml_version)
         external_versions.add(sklearn_version)
         for visitee in sub_components.values():
+            if isinstance(visitee, str):
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Same as above, why could this be a string?

File: openml/extensions/sklearn/extension.py [hunk 3]
```diff
@@ -769,21 +773,25 @@ def _get_external_version_string(
...
+                continue
             for external_version in visitee.external_version.split(','):
                 external_versions.add(external_version)
         return ','.join(list(sorted(external_versions)))
 
     def _check_multiple_occurence_of_component_in_flow(
         self,
         model: Any,
-        sub_components: Dict[str, OpenMLFlow],
+        sub_components: Dict[str, Any],
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Hm, why can this map to `Any` now?

File: openml/extensions/sklearn/extension.py [hunk 4]
```diff
@@ -769,21 +773,25 @@ def _get_external_version_string(
...
     ) -> None:
         to_visit_stack = []  # type: List[OpenMLFlow]
         to_visit_stack.extend(sub_components.values())
         known_sub_components = set()  # type: Set[str]
         while len(to_visit_stack) > 0:
             visitee = to_visit_stack.pop()
-            if visitee.name in known_sub_components:
+            if isinstance(visitee, str):
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Same as above.

File: openml/extensions/sklearn/extension.py [hunk 5]
```diff
@@ -796,7 +804,7 @@ def _extract_information_from_model(
     ) -> Tuple[
         'OrderedDict[str, Optional[str]]',
         'OrderedDict[str, Optional[Dict]]',
-        'OrderedDict[str, OpenMLFlow]',
+        'OrderedDict[str, Any]',
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Same as above.

File: openml/extensions/sklearn/extension.py [hunk 6]
```diff
@@ -860,7 +868,16 @@ def flatten_all(list_):
                         # length 3 is for ColumnTransformer
                         msg = 'Length of tuple does not match assumptions'
                         raise ValueError(msg)
-                    if not isinstance(sub_component, (OpenMLFlow, type(None))):
+
+                    if isinstance(sub_component, str):
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: This makes sense, but could you please add an explanation of which scikit-learn model these support?

Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: I shall add the explanations based on the example I know. However, as I mentioned in the PR comment, I am not sure if these changes/assumptions can generalize to ALL such sklearn models.
I don't think we need to handle cases wherein sklearn cannot process a user input, that is, a str in place of an estimator. For a `ColumnTransform` sklearn handles the keywords _drop_ and _passthrough_ by treating them as estimators that are not evaluated. Hence, I am explicitly handling this case.
Should I forage the sklearn docs a bit more to see if any such models exist?

Comment (mfeurer) [reply openml/extensions/sklearn/extension.py]: I don't know any other examples myself, so testing the example from the issue plus maybe one or two examples from the scikit-learn docs should be sufficient. I don't think you need to search for other scikit-learn models which allow strings, though. Maybe @amueller has some additional input?

File: openml/extensions/sklearn/extension.py (hunk 3 above)
Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: The annotations used earlier expected the _subcomponent_ to be OpenML, however, we now need to handle _str_ too. I had the type annotations as `Dict[str, Union[str, OpenMLFlow]]`. However, flake8 was not able to decipher the rest of the if...else bodies and was raising potential errors, which I am handling with exactly those if...else cases. 
I agree that having *Any* is loosening the type checks but I wanted to see if these changes affect test cases on Travis, so I went ahead with Any in this draft PR since I was unable to think of a better solution.

Comment (mfeurer) [reply openml/extensions/sklearn/extension.py]: Could you please paste the error you got for the annotation `Dict[str, Union[str, OpenMLFlow]]`?

Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: There were a dozen or more :neutral_face: 
One of them, for instance, was for [this](https://github.com/openml/openml-python/blob/78b163e8a3f7fe2c977a26912f7a7a4ae2742f2a/openml/extensions/sklearn/extension.py#L700).
_str has got no 'name' attribute'_

File: openml/extensions/sklearn/extension.py (hunk 6 above)
Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: Sure, will do so.

File: openml/extensions/sklearn/extension.py (hunk 3 above)
Comment (mfeurer) [reply openml/extensions/sklearn/extension.py]: I think the solution is to use an if/elif/else statement here, where the access to `name` is after a check if the instance is of type `OpenMLFlow`. I'm not really happy about this as it bloat up the code, but it will allow us to keep the annotations as meaningful as possible.

Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: Though test cases passed locally, it might fail on travis since I still get these two flake issues.

`openml/extensions/sklearn/extension.py:698: error: Item "str" of "Union[OpenMLFlow, str, None]" has no attribute "name"`
`openml/extensions/sklearn/extension.py:698: error: Item "None" of "Union[OpenMLFlow, str, None]" has no attribute "name"`

I am unable to think of how I can bypass how flake is parsing the if-conditions.
It doesn't make sense, for multiple types.

Comment (mfeurer) [reply openml/extensions/sklearn/extension.py]: That's annoying. You could try [cast](https://docs.python.org/3/library/typing.html#typing.cast).

File: openml/extensions/sklearn/extension.py [hunk 7]
```diff
@@ -782,15 +785,18 @@ def _get_external_version_string(
...
+        to_visit_stack = []  # type: List[Union[OpenMLFlow, str, None]]
         to_visit_stack.extend(sub_components.values())
         known_sub_components = set()  # type: Set[str]
+
         while len(to_visit_stack) > 0:
             visitee = to_visit_stack.pop()
-            if isinstance(visitee, str):
+            if isinstance(visitee, str):  # 'drop', 'passthrough' can be passed as estimators
                 known_sub_components.add(visitee)
+            elif visitee is None:  # a None step can be included in a Pipeline
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Can we maybe simplify the code and only support `passthrough` instead of `None`? It appears to me that `passthrough` can be used in the same places and is much clearer to handle. Maybe raise an error message if a user puts in a `None` step?

Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: On checking further, [FeatureUnion supports _drop_ and _None_](https://github.com/openml/openml-python/blob/cbf0d6a749a4f23245f15668a548f235c61b82f6/tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py#L678) while [Pipeline supports _passthrough_ and _None_](https://github.com/openml/openml-python/blob/cbf0d6a749a4f23245f15668a548f235c61b82f6/tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py#L1804).

Should we still _not_ support None and raise an Exception along the lines of: 
`None not allowed, please use 'drop' or 'passthrough'`?

However, when creating the sklearn Pipeline object, using _drop_ flags an error.
Whereas for FeatureUnion, using _passthrough_ is not flagged until runtime, i.e., calling _fit_.

Not sure what would be the best design here, especially wrt the Exception message, if we do not allow None.

**Comment by Neeratyoy** [comment]
@mfeurer one of the old unit tests that I had to change `test_serialize_feature_union` fails for versions `0.18.2` and `0.19.2`.
It appears that the manner in which we are serializing, the behaviour of one of the intermediate stages have changed since these two versions. To be more specific, the parameter metadata extraction.

Should I then try to change the code to see how this can be handled or skip that particular unit test?

File: tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py [hunk 8]
```diff
@@ -1804,7 +1804,6 @@ def test_run_on_model_with_empty_steps(self):
         clf = sklearn.pipeline.Pipeline([
             ('dummystep', 'passthrough'),  # adding 'passthrough' as an estimator
             ('prep', clf),
-            ('variancethreshold', None),  # adding 'None' as an estimator
```
Comment (mfeurer) [inline tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py]: Is there a reason you dropped this and didn't replace it by `drop`?

File: openml/extensions/sklearn/extension.py [hunk 9]
```diff
@@ -883,7 +879,12 @@ def flatten_all(list_):
                             raise ValueError(msg)
                         else:
                             pass
-                    elif not isinstance(sub_component, (OpenMLFlow, type(None))):
+                    elif isinstance(sub_component, type(None)):
+                        msg = 'Cannot serialize objects of None type. Please replace with a '\
```
Comment (mfeurer) [inline openml/extensions/sklearn/extension.py]: Could you please add a unit test that serializing `None` indeed fails? Also, could you please phrase it as `Please use a valid placeholder for None`?

**Comment by mfeurer** [comment]
> Should I then try to change the code to see how this can be handled or skip that particular unit test?

You can skip that unit test.

File: tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py (hunk 8 above)
Comment (Neeratyoy) [reply tests/test_extensions/test_sklearn_extension/test_sklearn_extension.py]: Yeah, since Pipeline supports only _None_ and _passthrough_.

File: openml/extensions/sklearn/extension.py (hunk 9 above)
Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: Done.

File: openml/extensions/sklearn/extension.py (hunk 3 above)
Comment (Neeratyoy) [reply openml/extensions/sklearn/extension.py]: This was fixed.
